index.bs

```diff
@@ -2943,6 +3027,60 @@ and an optional [=boolean=] |isNoised| (default false):
 1. Run [=obtain and deliver an aggregatable debug report on source registration=]
     with |dataType|, |source|, and |isNoised|.
 
+To <dfn>find sources with common destinations and reporting origin</dfn> given an [=attribution source=] |pendingSource|:
+1. Let |matchingSources| be a new [=list=].
+1. [=set/iterate|For each=] |source| of the user agent's [=attribution source cache=]:
+    1. Let |commonDestinations| be the [=set/intersection=] of |source|'s [=attribution source/attribution destinations=] and |pendingSource|'s [=attribution source/attribution destinations=].
+    1. If |commonDestinations| [=set/is empty=], [=iteration/continue=].
+    1. If |source|'s [=attribution source/reporting origin=] and |pendingSource|'s [=attribution source/reporting origin=] are not [=same origin=], [=iteration/continue=].
+    1. [=list/Append=] |source| to |matchingSources|.
+1. Return |matchingSources|.
+
+To <dfn>remove fake event reports for source</dfn> given an [=attribution source=] |source| and a [=moment=] |minTriggerTime|:
+1. If |source|'s [=attribution source/randomized response=] is null, return.
+1. [=set/Remove=] all elements from the [=event-level report cache=] whose [=event-level report/trigger time=] is greater than or equal to |minTriggerTime| and whose [=event-level report/source identifier=] is equal to |source|'s [=attribution source/source identifier=].
+
+To <dfn>remove sources with different max event states</dfn> given an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/max event states=] is null, return.
+1. Let |matchingSources| be the result of running [=find sources with common destinations and reporting origin=] with |pendingSource|.
+1. [=list/iterate|For each=] |source| of |matchingSources|:
+    1. If |source|'s [=attribution source/max event states=] is null or equals |pendingSource|'s [=attribution source/max event states=], [=iteration/continue=].
+    1. [=Remove fake event reports for source=] with |source| and |pendingSource|'s [=attribution source/source time=].
+    1. [=set/Remove=] |source| from the [=attribution source cache=].
+
+To <dfn>remove sources with lower attribution scope limit from cache</dfn> given an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/attribution scope limit=] is null, return.
+1. Let |matchingSources| be the result of running [=find sources with common destinations and reporting origin=] with |pendingSource|.
+1. [=list/iterate|For each=] |source| of |matchingSources|:
+    1. If |source|'s [=attribution source/attribution scope limit=] is null or greater than or equal to |pendingSource|'s [=attribution source/attribution scope limit=], [=iteration/continue=].
+    1. [=Remove fake event reports for source=] with |source| and |pendingSource|'s [=attribution source/source time=].
+    1. [=set/Remove=] |source| from the [=attribution source cache=].
```

Comment:
What do you think about combining all 3 of these algorithms, since we're really just iterating over sources with a particular reporting origin and processing them in a common way iff the source has at least one destination in common with the new source and either the scope check applies or the state check applies (or both).

Reply:
great point. Done.